File: frontend/src/intercom-widget.ts
import { LitElement } from 'lit';
import { property } from 'lit/decorators';

declare global {
    interface Window {
        Intercom: any;
    }
}

class IntercomWidget extends LitElement {

    @property({ type: String }) appId = "";
    @property({ type: String }) email = "";
    @property({ type: String }) userId = "";


    connectedCallback() {
        super.connectedCallback();
        if (this.appId) {
            window.Intercom('boot', {
                app_id: this.appId,
                email: this.email,
                user_id: this.userId
            });
        }
    }
    render() {
        return null;
    }

    show() {
        window.Intercom('show');
    }

    shutdown() {
        window.Intercom('shutdown');
    }

    update(attributes: object) {
        window.Intercom('update', attributes);
    }

    hide() {
        window.Intercom('hide');
    }

    showMessages() {
        window.Intercom('showMessages');
    }

    showNewMessage(preFilledContent?: string) {
        window.Intercom('showNewMessage', preFilledContent);
    }

    trackEvent(eventName: string, metadata: object) {
        window.Intercom('trackEvent', eventName, metadata);
    }
}

customElements.define('intercom-widget', IntercomWidget);

File: src/main/java/io/jmix/intercom/view/component/IntercomWidget.java
package io.jmix.intercom.view.component;
import com.vaadin.flow.component.JsonSerializable;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.component.littemplate.LitTemplate;
import elemental.json.JsonValue;

import java.io.Serializable;

@Tag("intercom-widget")
@JsModule("./src/intercom-widget.ts")
public class IntercomWidget extends LitTemplate {

    public IntercomWidget(String appId, String email, String userId) {
        getElement().setProperty("appId", appId);
        getElement().setProperty("email", email);
        getElement().setProperty("userId", userId);
    }

    public void show() {
        getElement().callJsFunction("show");
    }

    public void shutdown() {
        getElement().callJsFunction("shutdown");
    }

    public void update(Serializable attributes) {
        getElement().callJsFunction("update", attributes);
    }

    public void hide() {
        getElement().callJsFunction("hide");
    }

    public void showMessages() {
        getElement().callJsFunction("showMessages");
    }

    public void showNewMessage(String preFilledContent) {
        getElement().callJsFunction("showNewMessage", preFilledContent);
    }

    public void trackEvent(String eventName, JsonValue metadata) {
        getElement().callJsFunction("trackEvent", eventName, metadata);
    }
}
